### ISSUE TITLE
Optimize index construction

### ISSUE THREAD

**Issue by rmcolq** [rmcolq/pandora#23]

**Comment by rmcolq** [comment]
Indexing the 23,000 genes takes 3 days 5 hours. Here are some excerpts from the log file for genes which took a long time
```
Wed Aug 29 15:09:20 2018 Sketch PRG GC00000205_10 which has 1000 nodes
Wed Aug 29 15:13:03 2018 Remove 'bad' edges from kmergraph

Wed Aug 29 15:51:42 2018 Sketch PRG GC00000027_6 which has 4088 nodes
Wed Aug 29 19:48:21 2018 Remove 'bad' edges from kmergraph

Wed Aug 29 22:07:03 2018 Sketch PRG GC00004221 which has 4572 nodes
Thu Aug 30 19:42:04 2018 Remove 'bad' edges from kmergraph

Thu Aug 30 19:43:45 2018 Sketch PRG GC00002479 which has 1287 nodes
Thu Aug 30 20:02:25 2018 Remove 'bad' edges from kmergraph

Thu Aug 30 20:54:11 2018 Sketch PRG GC00000222_4 which has 1537 nodes
Thu Aug 30 21:15:27 2018 Remove 'bad' edges from kmergraph

Thu Aug 30 21:15:42 2018 Sketch PRG GC00002914 which has 2470 nodes
Thu Aug 30 21:33:54 2018 Remove 'bad' edges from kmergraph

Thu Aug 30 22:24:38 2018 Sketch PRG GC00002737 which has 1978 nodes
Thu Aug 30 23:12:41 2018 Remove 'bad' edges from kmergraph

Fri Aug 31 23:56:52 2018 Sketch PRG GC00003373_r1_1 which has 4666 nodes
Sat Sep  1 01:22:40 2018 Remove 'bad' edges from kmergraph
```
Compared to fast times with medium numbers of nodes
```
Wed Aug 29 15:06:42 2018 Sketch PRG GC00007068_2 which has 221 nodes
Wed Aug 29 15:06:45 2018 Remove 'bad' edges from kmergraph

Wed Aug 29 15:06:46 2018 Sketch PRG GC00000085_5 which has 288 nodes
Wed Aug 29 15:06:51 2018 Remove 'bad' edges from kmergraph

Wed Aug 29 15:06:59 2018 Sketch PRG GC00001278 which has 656 nodes
Wed Aug 29 15:08:26 2018 Remove 'bad' edges from kmergraph

Wed Aug 29 15:08:47 2018 Sketch PRG GC00002569 which has 500 nodes
Wed Aug 29 15:09:20 2018 Remove 'bad' edges from kmergraph

Wed Aug 29 15:13:04 2018 Sketch PRG GC00000602_2 which has 369 nodes
Wed Aug 29 15:13:18 2018 Remove 'bad' edges from kmergraph
```

**Comment by rmcolq** [comment]
PRGs in txt format so github would allow them:
[fast_prg.txt](https://github.com/rmcolq/pandora/files/2415760/fast_prg.txt)
[slow_prg.txt](https://github.com/rmcolq/pandora/files/2415761/slow_prg.txt)

**Comment by rmcolq** [comment]
These were indexed using
```
singularity exec /nfs/leia/research/iqbal/rmcolq/projects/pandora_variation/analysis/loman_k12/work/singularity/rmcolq-pandora-pandora.img pandora index -w 14 -k 15 ecoli_pangenome_PRG_290818.fa
```

**Comment by rffrancon** [comment]
I intend on profiling using valgrind. I expect that a singularity container would get in the way. Which pandora commit should I use in profiling?

**Comment by rmcolq** [comment]
Even if you use the current commit on dev, it would be fine, as the problems are only when running a function not in anyway involved in the `index` step.

**Comment by rffrancon** [comment]
That's good to know. I will work from this commit: 6bf320a81ae7cf39558782cd1afb07db46731021
Please let me know if that's not ok.

**Comment by rffrancon** [comment]
Baseline profiling results:
[profiling_results.zip](https://github.com/rmcolq/pandora/files/2434338/profiling_results.zip)

The above zip also includes profiling results after basic refactoring. The results suggest ~15% improvement. Open pull request: https://github.com/rmcolq/pandora/pull/37

**Comment by rffrancon** [comment]
![baseline](https://user-images.githubusercontent.com/5958289/46296398-34691880-c592-11e8-8060-4b0d00a162cf.png)

**Comment by rffrancon** [comment]
Following commit 843f959837a0597a5f77898a361fd460a2ccc215 (memorization for `Interval::get_end()`):

![commit_843f959837a0597a5f77898a361fd460a2ccc215](https://user-images.githubusercontent.com/5958289/46361785-62b82800-c666-11e8-8c02-205259bf571e.png)

**Comment by rmcolq** [comment]
NB, Zam's suggestion of parallelising is possible if genes are passed their own index, and an index merge function is added

**Comment by rffrancon** [comment]
@rmcolq I thought we were using the `thesis` tag for issues that had to be completed before the thesis could be finished. Is that the case for this issue?

**Comment by leoisl** [comment]
hello,

TL;DR: 8.8x speed up on [fast_prg.txt ](https://github.com/rmcolq/pandora/files/2415760/fast_prg.txt) by using IntervalTree (https://github.com/lh3/cgranges) for fast interval overlap queries.

https://github.com/rmcolq/pandora/issues/23#issuecomment-426334840  shows that `LocalPRG::nodes_along_path()` is the bottleneck. This part of the code makes several interval overlap queries, which takes linear time in the worst case (for each interval in a kmer path, we get the PRG nodes overlapping it). I added an IntervalTree (https://github.com/lh3/cgranges) index in LocalGraph, which is very fast for such queries (also had to add two other std::map to deal with intervals of length 0, but this is a detail). Interval queries now take log time.

(Very simple) benchmarking on [fast_prg.txt ](https://github.com/rmcolq/pandora/files/2415760/fast_prg.txt) with this improvement (https://github.com/leoisl/pandora/commit/7e35cf9761b2d568acfe43fc21b0e2464aeae9f3):
```
root@leandro-ThinkPad-X1-Carbon-6th:/pandora/data_issue_23# /usr/bin/time --verbose ../build_debug/pandora index fast_prg.txt 
[2019-05-13 21:37:33.372016] [0x00007fd3f6af7740] [debug]   prgfile: fast_prg.txt
[2019-05-13 21:37:33.411412] [0x00007fd3f6af7740] [info]    Sketch PRG GC00007068_2 which has 221 nodes
[2019-05-13 21:37:41.708654] [0x00007fd3f6af7740] [info]    Sketch PRG GC00000085_5 which has 288 nodes
[2019-05-13 21:37:51.215872] [0x00007fd3f6af7740] [info]    Sketch PRG GC00002569 which has 500 nodes
	Command being timed: "../build_debug/pandora index fast_prg.txt"
...
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:54.64
```

Versus with no IntervalTree (https://github.com/rmcolq/pandora/commit/7adf63ce60d28a000f7b6c850f4a5ebbbf2dd031) - had to remove `-O2` from https://github.com/rmcolq/pandora/blob/master/CMakeLists.txt#L24 for fairness - both compiled in debug mode (`-DCMAKE_BUILD_TYPE=DEBUG`):
```
root@leandro-ThinkPad-X1-Carbon-6th:/pandora/data_issue_23# /usr/bin/time --verbose ../build_debug/pandora index fast_prg.txt
[2019-05-13 21:38:46.894428] [0x00007fc273bf0740] [debug]   prgfile: fast_prg.txt
[2019-05-13 21:38:46.926415] [0x00007fc273bf0740] [info]    Sketch PRG GC00007068_2 which has 221 nodes
[2019-05-13 21:39:28.113321] [0x00007fc273bf0740] [info]    Sketch PRG GC00000085_5 which has 288 nodes
[2019-05-13 21:40:30.195688] [0x00007fc273bf0740] [info]    Sketch PRG GC00002569 which has 500 nodes
	Command being timed: "../build_debug/pandora index fast_prg.txt"
...
	Elapsed (wall clock) time (h:mm:ss or m:ss): 7:58.55
```
So it is a 478÷54=8.8x speed up in this example. I bet that in more complex cases (PRGs with a lot of nodes), the time difference will be larger, as log and linear functions will have larger gap.

All tests pass with ctest and I diffed both outputs and both versions produce the same index, but I would be sceptical if this implementation is indeed fine - I just know a part of pandora index, not the whole codebase yet...

I will be able to get an updated profiling only tomorrow, as callgrind is slooooow...

Can discuss tomorrow if you wish.

Hoping to improve further and then multithread the index building.

**Comment by rmcolq** [comment]
Awesome!! I look forward to hearing how this improves the slow ones when we know!

**Comment by leoisl** [comment]
Updated profiling:

![callgrind 7e35cf fast_prg](https://user-images.githubusercontent.com/8397967/57660255-9801fc80-75dd-11e9-82ad-56bb81b8f3d8.jpg)

**Comment by leoisl** [comment]
Updates:
* Memoization of LocalPRG::nodes_along_path(): https://github.com/leoisl/pandora/commit/f5aee2aa4c7098d7d03a6875da6c52fc22027f81
* Coase-grained multi-threading added (1 thread/gene): https://github.com/leoisl/pandora/commit/e9e5d4d7177058520390b1dfea07c682ba06e75d and https://github.com/leoisl/pandora/commit/c208a39d45a3c93dd9d96b60b5cb88c8722c35b0
* LocalPRG::minimizer_sketch() performance improved: https://github.com/leoisl/pandora/commit/946c68f7672903560abf222dc1db28472097b46f


Updated profiling:
![callgrind 946c68 fast_prg](https://user-images.githubusercontent.com/8397967/57764348-27e99880-76fb-11e9-8c4e-a9056ebb7089.jpg)

**Comment by leoisl** [comment]
Log time memoization of LocalPRG::nodes_along_path() taking 25% of CPU instructions, you were right, gotta go constant time...

**Comment by rmcolq** [comment]
Good luck...

**Comment by leoisl** [comment]
Constant-time memoization done in https://github.com/leoisl/pandora/commit/27dec3b45d4780c39444dbff2727bf906c981e12 .

I don't think there are things that are easy to optimize now, as the update profiling shows:

![callgrind 27dec3 fast_prg](https://user-images.githubusercontent.com/8397967/57790224-3901cc00-7732-11e9-9b8c-5cc4ef7cea5f.jpg)

**Comment by iqbal-lab** [comment]
what is the difference in performance (as opposed to the change in proportion of time/profile)? have i missed that in the above figure? sorry

**Comment by leoisl** [comment]
I don't think I understood your question... Valgrind callgrind allows you to do instruction profiling, not time profiling (https://stackoverflow.com/questions/6663614/use-valgrind-to-know-timein-seconds-spent-in-each-function). Sorry if this was not what you meant, could you explain better please?

Anyway, follows some preliminary time profiling:

Comparing the code performance before and after these optimizations with build type on Release mode:

fast_prg.txt runtimes:
* https://github.com/leoisl/pandora/commit/27dec3b45d4780c39444dbff2727bf906c981e12 (1 thread): 7s
* https://github.com/rmcolq/pandora/commit/7adf63ce60d28a000f7b6c850f4a5ebbbf2dd031 : 59s
* **built indexes are the same**

slow_prg.txt (genes GC00002914, GC00004221 and GC00003373_r1_1) runtimes:
* https://github.com/leoisl/pandora/commit/27dec3b45d4780c39444dbff2727bf906c981e12 (1 thread): 18m 28s
* https://github.com/rmcolq/pandora/commit/7adf63ce60d28a000f7b6c850f4a5ebbbf2dd031 : >=1 day
* **I computed https://github.com/rmcolq/pandora/commit/7adf63ce60d28a000f7b6c850f4a5ebbbf2dd031 runtime by looking at https://github.com/rmcolq/pandora/issues/23#issuecomment-424379266 and calculating the time from one gene to another**:
```
GC00002914 = 1h9m
GC00004221 = 21h35m
GC00003373_r1_1 = at least 1h26m (can't know for sure since it is the last gene and we don't have the time taken by the remove bad edges part)
```
* **runtimes for this test case were measured on different machines (and my machine has an amazing i7 8th gen, it is probably faster than the processor that ran https://github.com/rmcolq/pandora/issues/23#issuecomment-424379266)**
* **cannot verify if the built indexes are the same, as we don't have the built index for slow_prg.txt on https://github.com/rmcolq/pandora/commit/7adf63ce60d28a000f7b6c850f4a5ebbbf2dd031**

I will proceed by running https://github.com/leoisl/pandora/commit/27dec3b45d4780c39444dbff2727bf906c981e12 on large datasets given to me by Michael and Rachel. I think I am fine with the performance now anyway, since it does not seem trivial to improve. I think what can be addressed next is multi-threading within the gene. The bottleneck of this test was GC00004221, which took 17m of the 18 total minutes. If we multi-thread within the gene, hopefully we can take this down to ~2m if using 8 cores. Do you agree with this? If not, what should be addressed next?

**Comment by leoisl** [comment]
I guess improving test coverage on the classes concerning pandora index with https://github.com/rmcolq/pandora/wiki/Generate-code-coverage-reports has more priority than multithreading within gene? Better to make sure all these performance improvements aren't introducing new bugs, no?

**Comment by iqbal-lab** [comment]
This is probably also a good time to raise the issue of how we are going to do code review. easier verbally tomorrow.

**Comment by mbhall88** [comment]
> I will proceed by running [leoisl@27dec3b](https://github.com/leoisl/pandora/commit/27dec3b45d4780c39444dbff2727bf906c981e12) on large datasets given to me by Michael and Rachel. I think I am fine with the performance now anyway, since it does not seem trivial to improve. I think what can be addressed next is multi-threading within the gene. The bottleneck of this test was GC00004221, which took 17m of the 18 total minutes. If we multi-thread within the gene, hopefully we can take this down to ~2m if using 8 cores. Do you agree with this? If not, what should be addressed next?

In previous timings I have done the longest gene to index was `GC00004221` at approx. 28hrs. **So given that is now taking 17mins that is approx. 99x speedup!**

**Comment by leoisl** [comment]
Profiling/testing https://github.com/leoisl/pandora/commit/973e607687b77ce542f46bd1529cc8b1eba0261c on `/hps/nobackup/iqbal/rmcolq/projects/panrg/ecoli/k15.w14/ecoli_pangenome_PRG_050319.fa` with `/usr/bin/time`:
* Build type was release (`cmake -DCMAKE_BUILD_TYPE=Release ..`)
* Command-line: `/usr/bin/time --verbose /pandora/build_release/pandora index ecoli_pangenome_PRG_050319.fa -t 8`
* Time report:
```
	Command being timed: "/pandora/build_release/pandora index ecoli_pangenome_PRG_050319.fa -t 8"
	User time (seconds): 35030.10
	System time (seconds): 61.53
	Percent of CPU this job got: 783%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 1:14:41
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 13488736
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 88
	Minor (reclaiming a frame) page faults: 3372386
	Voluntary context switches: 10956
	Involuntary context switches: 2098877
	Swaps: 0
	File system inputs: 23728
	File system outputs: 1871968
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```
* Important points:
  * Percent of CPU this job got: 783%: **The best we can get here is 800%, since the command was run with 8 threads (-t 8), so multithreading seems to work fine.**
  * Elapsed (wall clock) time (h:mm:ss or m:ss): 1:14:41: **This is the time that matters (how much we really waited): ~1h15min**
  * Maximum resident set size (kbytes): 13488736 **This is ~12.8 GB max used RAM**

* Sanity check:
  * Comparing the indexes with https://github.com/leoisl/pandora/blob/973e607687b77ce542f46bd1529cc8b1eba0261c/scripts/compare_indexes/compare_indexes.sh:
```
root@leandro-ThinkPad-X1-Carbon-6th:/pandora/scripts/compare_indexes# bash compare_indexes.sh ../../build_release/bin/compare_indexes /datasets/data_issue_23_large_dataset/ecoli_pangenome_PRG_050319.fa.k15.w14.idx /datasets/data_issue_23_large_dataset/kmer_prgs/ /datasets/data_issue_23_large_dataset/testing_973e60/ecoli_pangenome_PRG_050319.fa.k15.w14.idx /datasets/data_issue_23_large_dataset/testing_973e60/kmer_prgs/
+ ../../build_release/bin/compare_indexes /datasets/data_issue_23_large_dataset/ecoli_pangenome_PRG_050319.fa.k15.w14.idx /datasets/data_issue_23_large_dataset/testing_973e60/ecoli_pangenome_PRG_050319.fa.k15.w14.idx
[2019-05-16 12:17:54.840142] [0x00007fa700d5cf80] [debug]   Loading index
[2019-05-16 12:17:54.840214] [0x00007fa700d5cf80] [debug]   File is /datasets/data_issue_23_large_dataset/ecoli_pangenome_PRG_050319.fa.k15.w14.idx
[2019-05-16 12:18:01.616964] [0x00007fa700d5cf80] [debug]   Finished loading file. Index now contains 4685539 entries
[2019-05-16 12:18:01.616994] [0x00007fa700d5cf80] [debug]   Loading index
[2019-05-16 12:18:01.616997] [0x00007fa700d5cf80] [debug]   File is /datasets/data_issue_23_large_dataset/testing_973e60/ecoli_pangenome_PRG_050319.fa.k15.w14.idx
[2019-05-16 12:18:09.336882] [0x00007fa700d5cf80] [debug]   Finished loading file. Index now contains 4685539 entries
Indexes are EQUAL
+ diff -rq /datasets/data_issue_23_large_dataset/kmer_prgs/ /datasets/data_issue_23_large_dataset/testing_973e60/kmer_prgs/
+ echo 'If you are seeing this message, then the given indexes are identical!'
If you are seeing this message, then the given indexes are identical!
```

So https://github.com/leoisl/pandora/commit/973e607687b77ce542f46bd1529cc8b1eba0261c produced the same index as the previous version for a PRG of 37426 genes. I think we still need to ensure that no bugs were introduced, with unit tests covering 100% of the `pandora index` classes, but this is a good sign at least that it seems to be working. Be aware that the other modules (`map`, `compare`, etc) might break because I made changes in some classes that are used in several other modules, but just tested well `index`.

Final questions:
1. Is this time performance ok? I don't know what you got using nextflow/snakemake on this PRG. Improvements in time performance can still be done, multithreading can be done better, etc...
2. Is memory usage ok? Should we try to reduce it?

**Comment by mbhall88** [comment]
> Final questions:
> 
>     1. Is this time performance ok? I don't know what you got using nextflow/snakemake on this PRG. Improvements in time performance can still be done, multithreading can be done better, etc...

Well given this was taking ~7 days to index previously, I would say this is more than ok! 🚀 

>     2. Is memory usage ok? Should we try to reduce it?

See what Zam and Rachel think, but for now, ~13GB (peak) seems more than reasonable to me.

**Comment by iqbal-lab** [comment]
i think this is awesome. Just over an hour using 8 threads is great, and 13Gb of RAM is fine by me. ok @rmcolq ?

**Comment by leoisl** [comment]
> See what Zam and Rachel think, but for now, ~13GB (peak) seems more than reasonable to me. i.e I think there are more pressing things to work on than the `index` memory usage.

I see. I also forgot to mention that memory increases with nb of threads. I would not be surprised if it takes ~26GB with 16 threads... If it is ran with a lot of threads, RAM might become an issue. But I agree that this should not have high priority.

I guess cheap addition of new sequences (discovered with denovo) is the highest priority for `index` now?

**Comment by iqbal-lab** [comment]
i mean - at this speed, no real need for nextmake or snakeflow

**Comment by iqbal-lab** [comment]
yes, cheap addition good, but you have now made it so fast, even a complete rebuild is bearable. since you've partially implemented cheap append, it is still the right thing to do of course, but this is great performance

**Comment by leoisl** [comment]
I will do a second benchmark on yoda with 16 threads to see how it goes. Would you have another, larger PRG?

**Comment by rmcolq** [comment]
That is the largest and longest PRG which I have at the moment. All sounds really great! And I agree with the others

**Comment by leoisl** [comment]
Running on yoda:
```
Command being timed: "/hps/nobackup/iqbal/leandro/pandora/build_release/pandora index /hps/nobackup/iqbal/leandro/pandora_data/ecoli_pangenome_PRG_050319.fa -t 16"
	User time (seconds): 58560.62
	System time (seconds): 251.93
	Percent of CPU this job got: 935%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 1:44:49
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 13522272
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 7
	Minor (reclaiming a frame) page faults: 20465413
	Voluntary context switches: 1359193
	Involuntary context switches: 2796731
	Swaps: 0
	File system inputs: 4552
	File system outputs: 1880888
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

Issues:
* Percent of CPU this job got: 935% **I asked for 16 threads, but on average only 9.3 executed concurrently. My guess is that synchronization must be improved... Maybe it is not very useful to put more than 10 threads in this version of `pandora index`... to test empirically yet**
* Elapsed (wall clock) time (h:mm:ss or m:ss): 1:44:49 **Time was longer than using 8 cores of my laptop, but still fine**
* Maximum resident set size (kbytes): 13522272 **RAM was fine ~13GB**

I could verify that the index was built correctly.

We seem done here, I am closing, feel free to reopen if needed.

**Comment by leoisl** [comment]
Reopening, I was profiling `pandora index` with the wrong build type. I was profiling a debug build, but what we have to profile is a release build with debug flags.

I created a wiki page to help us remember which build type should be used accordingly to what we want to do: https://github.com/rmcolq/pandora/wiki/How-to-build-pandora-for-debugging,-profiling-and-release

I will update this with the new profiling.

**Comment by leoisl** [comment]
Updated profile with profiling build on fast.prg:

[callgrind.afbb09d.fast_prg.out.txt](https://github.com/rmcolq/pandora/files/3191456/callgrind.afbb09d.fast_prg.out.txt)

![callgrind afbb09d fast_prg](https://user-images.githubusercontent.com/8397967/57928056-9d926780-78a7-11e9-9cb2-5f4dc969abe6.jpg)

Getting a profile on slow_prg also.

**Comment by leoisl** [comment]
TL;DR: Solved performance issue due to heavy deallocation and reallocation of `vector<prg::Path>` in `LocalPRG::minimizer_sketch()` and in `LocalPRG_shift()` (see https://github.com/leoisl/pandora/commit/5508d2b0e144d6fdf1eb9fadf670dc5db8043c3e). Performance improved, and luckily and very importantly, this happened also to be what was holding multi-threading performance back (I'd never guess this...). On a previous commit (see https://github.com/rmcolq/pandora/issues/23#issuecomment-493123670), running on yoda and asking for 16 threads, only 9.3 threads executed on average and the runtime was 1h 44min. **With the performance fixes on https://github.com/leoisl/pandora/commit/5508d2b0e144d6fdf1eb9fadf670dc5db8043c3e, running on yoda also asking for the same 16 threads gets instead 15.4 threads executing on average, and the runtime is 35min.** I am closing this as I am afraid I am unable to improve performance further easily, and this runtime seems fine.

Details follow:

Benchmarking https://github.com/leoisl/pandora/commit/5508d2b0e144d6fdf1eb9fadf670dc5db8043c3e  on `/hps/nobackup/iqbal/rmcolq/projects/panrg/ecoli/k15.w14/ecoli_pangenome_PRG_050319.fa` on yoda:
```
	Command being timed: "/hps/nobackup/iqbal/leandro/pandora/build_release/pandora index /hps/nobackup/iqbal/leandro/pandora_data/ecoli_pangenome_PRG_050319.fa -t 16"
        [...]
	Percent of CPU this job got: 1540%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 35:16.37
        [...]
	Maximum resident set size (kbytes): 13510072
```

**Multi-threading now seems fine, it was not a too-much-synchronization issue as I thought. We asked for 16 threads, and 15.4 executed on average. This plus the improved performance dropped the runtime to 35 minutes. Maximum RAM usage is ~12.8 GB, the same as running with 8 threads (https://github.com/rmcolq/pandora/issues/23#issuecomment-493047296), so it seems that using more threads do not really impact a lot on RAM usage. I verified that the index was built correctly.**

Additional resource usage info from the cluster:
```
Resource usage summary:

    CPU time :                                   32597.16 sec.
    Max Memory :                                 13195 MB
    Average Memory :                             7155.40 MB
    Total Requested Memory :                     20000.00 MB
    Delta Memory :                               6805.00 MB
    Max Swap :                                   13788 MB
    Max Processes :                              4
    Max Threads :                                20
    Run time :                                   2117 sec.
    Turnaround time :                            2117 sec.
```

Updated profiling using profiling build:
[callgrind.5508d2b.profiling_build.fast_prg.txt](https://github.com/rmcolq/pandora/files/3194872/callgrind.5508d2b.profiling_build.fast_prg.txt)
![callgrind 5508d2b profiling_build fast_prg](https://user-images.githubusercontent.com/8397967/57979270-46110a80-7a13-11e9-881a-af5edf5e63b4.jpg)

Looking at the profiling, it seems hard to improve the performance now. Also, this performance seems fine now that multi-threading works well. If performance needs to be improved in the future, I think first thing to do is multi-threading within gene. If this is not enough, then I guess we would need to change the minimizer sketch algorithm, but this should be seen just as the last option. The algorithm is efficient already, and most importantly, we know it is stable and produce correct indexes. Changing it is too dangerous, and should be considered only if no more options are available. **Anyway, any path for performance improvement from here is very hard to implement, so I am closing this issue.**

Just for completeness: updated profiling using debug build (we should not profile using this, but it is a lot easier to understand as functions don't get inlined):
[callgrind.5508d2b.debug_build.fast_prg.txt](https://github.com/rmcolq/pandora/files/3194873/callgrind.5508d2b.debug_build.fast_prg.txt)
![callgrind 5508d2b debug_build fast_prg](https://user-images.githubusercontent.com/8397967/57979284-8c666980-7a13-11e9-98d6-6afc81a8a384.jpg)

**Comment by mbhall88** [comment]
Phenomenal work @leoisl . This is super exciting!!  

Probably a good time for a pull request on to dev?

**Comment by leoisl** [comment]
Hello!

Not sure, some of the modified code might not be covered by unit tests... Maybe #133 is required for a PR? I modified some classes/code that are also used in the other modules, but I just checked that `index` seems to be working fine (still only in one dataset... Would you happen to have another dataset so that we can compare both versions more comprehensively?), but I am unsure if `map` or `compare` for example broke. The unit tests all pass though...

Also, I think Rachel is running some last experiments for her PhD thesis, so it might be better to keep dev stable as we know it is now, and make the PR after she finishes the text?

**Comment by mbhall88** [comment]
What does the code coverage look like for the functions you were working on?

**Comment by leoisl** [comment]
It seems to be totally covered (see #133 for details).

Also, what I thought it would break in `map` (some additional data structures that I added and that were not serialized) seems to be fine since they are rebuilt on deserialization. So I guess everything is fine. I will just map all reads from `/nfs/leia/research/iqbal/rmcolq/data/cardio/illumina/H124200644.fq` on `/hps/nobackup/iqbal/rmcolq/projects/panrg/ecoli/k15.w14/ecoli_pangenome_PRG_050319.fa` and compare both mapping outputs just to verify if mapping is indeed not broken.

If this runs fine, I will make the PR!
